handlers/smshandler/mongodbstorage.go

```diff
@@ -0,0 +1,385 @@
+}
+
+func (ms *MongoStorage) User(userID types.HexBytes) (*UserData, error) {
+	ms.keysLock.RLock()
+	defer ms.keysLock.RUnlock()
+	user, err := ms.getUserData(userID)
+	if err != nil {
+		return nil, err
+	}
+	return user, nil
+}
+
+func (ms *MongoStorage) getUserData(userID types.HexBytes) (*UserData, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	result := ms.users.FindOne(ctx, bson.M{"_id": userID})
+	var user UserData
+	if err := result.Decode(&user); err != nil {
+		log.Warn(err)
+		return nil, ErrUserUnknown
+	}
+	return &user, nil
+}
+
+func (ms *MongoStorage) updateUser(user *UserData) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	_, err := ms.users.ReplaceOne(ctx, bson.M{"_id": user.UserID}, user)
+	if err != nil {
+		return fmt.Errorf("cannot update object: %w", err)
+	}
+	return nil
+}
+
+func (ms *MongoStorage) UpdateUser(udata *UserData) error {
+	ms.keysLock.RLock()
+	defer ms.keysLock.RUnlock()
+	return ms.updateUser(udata)
+}
+
+func (ms *MongoStorage) BelongsToElection(userID types.HexBytes,
+	electionID types.HexBytes) (bool, error) {
+	ms.keysLock.RLock()
+	defer ms.keysLock.RUnlock()
+	user, err := ms.getUserData(userID)
+	if err != nil {
+		return false, err
+	}
+	ei := user.FindElection(electionID)
+	return ei >= 0, nil
+}
+
+func (ms *MongoStorage) IncreaseAttempt(userID, electionID types.HexBytes) error {
+	ms.keysLock.Lock()
+	defer ms.keysLock.Unlock()
+
+	user, err := ms.getUserData(userID)
+	if err != nil {
+		return err
+	}
+
+	ei := user.FindElection(electionID)
+	if ei == -1 {
+		return ErrUserNotBelongsToElection
+	}
+	user.Elections[ei].RemainingAttempts++
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	_, err = ms.users.ReplaceOne(ctx, bson.M{"_id": userID}, user)
+	if err != nil {
+		return fmt.Errorf("cannot update object: %w", err)
+	}
+	return nil
+}
+
+func (ms *MongoStorage) NewAttempt(userID, electionID types.HexBytes,
+	challenge int, token *uuid.UUID) (*phonenumbers.PhoneNumber, error) {
+	ms.keysLock.Lock()
+	defer ms.keysLock.Unlock()
+
+	user, err := ms.getUserData(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	ei := user.FindElection(electionID)
+	if ei == -1 {
+		return nil, ErrUserNotBelongsToElection
+	}
+	if user.Elections[ei].Consumed {
+		return nil, ErrUserAlreadyVerified
+	}
+	if user.Elections[ei].RemainingAttempts < 1 {
+		return nil, ErrTooManyAttempts
+	}
+	user.Elections[ei].RemainingAttempts--
+	user.Elections[ei].AuthToken = token
+	user.Elections[ei].Challenge = challenge
+
+	if err := ms.updateUser(user); err != nil {
+		return nil, err
+	}
+
+	// Save the token as index for finding the userID
+	atindex := AuthTokenIndex{
+		AuthToken: token,
+		UserID:    user.UserID,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	_, err = ms.tokenIndex.InsertOne(ctx, atindex)
```

Comment:
maybe the error can be checked after returning so if there is this last error happening the function also returns nil like some lines above